src/convert.rs

```diff
@@ -0,0 +1,139 @@
+/// Converts spiral coordinates to and from cube (q,r,s) coordinates
+use crate::position::{ring, ring_offset};
+
+/// Convert spiral hex coordinate x to cube coords (q,r,s).
+/// See: https://www.redblobgames.com/grids/hexagons/
+/// for a definition of cube coords.
+fn spiral_to_cube(x: usize) -> Result<(i8, i8, i8), &'static str> {
+    // The origin is a special case: return (0,0,0)
+    if x == 0 {
+        return Ok((0, 0, 0));
+    }
+
+    // Find the ring number and ring-offset for this spiral
+    let ring_number = ring(x) as f32;
+    let ring_offset = ring_offset(ring_number as usize) as f32;
+
+    // Calculate q and r
+    let q = growing_trunc_tri(x as f32, ring_number, ring_offset, 0.0);
+    let r = growing_trunc_tri(x as f32, ring_number, ring_offset, 4.0);
+
+    // Could alternatively manually calculate s as:
+    // let s = growing_trunc_tri(x, ring_offset, p, ring_number, -4.0);
+    let s = -q - r;
+
+    Ok((q, r, s))
+}
+
+/// Calculate a spiral hex coordinate for an input (q,r,s) in cube coordinates
+fn cube_to_spiral(qrs: (i8, i8, i8)) -> Result<usize, &'static str> {
```

Comment:
it might make sense to either create a `struct Qrs` or to pass those as `q: i8, r: i8, s: i8`, as they get separated immediately